{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Converts a 3 base codon into an amino acid\n",
    "def codon_to_aa(codon_amino):\n",
    "    codon_dict = {'TTT':'Phe','TTC':'Phe','TTA':'Leu','TTG':'Leu','CTT':'Leu','CTC':'Leu','CTA':'Leu','CTG':'Leu','ATT':'Ile','ATC':'Ile','ATA':'Ile','ATG':'Met','GTT':'Val','GTC':'Val','GTA':'Val','GTG':'Val','TCT':'Ser','TCC':'Ser','TCA':'Ser','TCG':'Ser','CCT':'Pro','CCC':'Pro','CCA':'Pro','CCG':'Pro','ACT':'Thr','ACC':'Thr','ACA':'Thr','ACG':'Thr','GCT':'Ala','GCC':'Ala','GCA':'Ala','GCG':'Ala','TAT':'Tyr','TAC':'Tyr','TAA':'STOP','TAG':'STOP','CAT':'His','CAC':'His','CAA':'Gln','CAG':'Gln','AAT':'Asn','AAC':'Asn','AAA':'Lys','AAG':'Lys','GAT':'Asp','GAC':'Asp','GAA':'Glu','GAG':'Glu','TGT':'Cys','TGC':'Cys','TGA':'STOP','TGG':'Trp','CGT':'Arg','CGC':'Arg','CGA':'Arg','CGG':'Arg','AGT':'Ser','AGC':'Ser','AGA':'Arg','AGG':'Arg','GGT':'Gly','GGC':'Gly','GGA':'Gly','GGG':'Gly'}\n",
    "    \n",
    "    #match codon with amino acid in dictionary\n",
    "    for i in codon_dict:\n",
    "        if i == codon_amino:\n",
    "            codon_amino = codon_dict[i]\n",
    "    return codon_amino"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Open reading frame\n",
    "class ORF:\n",
    "    #initialization\n",
    "    def __init__(self, orf_num, orf_bps, bp_start, bp_stop, orf_seq):\n",
    "        self.orf_num = orf_num\n",
    "        self.orf_bps = orf_bps\n",
    "        self.bp_start = bp_start\n",
    "        self.bp_stop = bp_stop\n",
    "        self.orf_seq = orf_seq\n",
    "    \n",
    "    #print variables\n",
    "    def print_attributes(self):\n",
    "        print(\"Orf attributes:\")\n",
    "        print(f\"The orf number is: {self.orf_num}\")\n",
    "        print(f\"Basepairs: {self.orf_bps}\")\n",
    "        print(f\"Starting at {self.bp_start}, ending at {self.bp_stop}\\n\")\n",
    "    \n",
    "    #Converts list of codons into a Protein object\n",
    "    def transcribe(self):\n",
    "        aa_seq = []\n",
    "        \n",
    "        for codon in self.orf_seq:\n",
    "            amino = codon_to_aa(codon)#uses codon-amino dictionary to convert\n",
    "            aa_seq.append(amino)\n",
    "        \n",
    "        ascendence = Protein(self.orf_num,len(aa_seq),aa_seq) #protein object \n",
    "        return ascendence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Protein:\n",
    "    #initialization\n",
    "    def __init__(self, prot_num, aa_count, aa_seq):\n",
    "        self.prot_num = prot_num\n",
    "        self.aa_count = aa_count\n",
    "        self.aa_seq = aa_seq\n",
    "    #print variables\n",
    "    def print_attributes(self):\n",
    "        print(self.prot_num, self.aa_count, self.aa_seq)\n",
    "    def print_aa(self):\n",
    "        print(f\"The amino acids sequence for this protein is {self.aa_seq}\")\n",
    "    \n",
    "    #gives summary of how many of each aa in protein\n",
    "    def aa_summary(self):\n",
    "        sum_dict = {'Phe': 0, 'Leu': 0, 'Ile': 0, 'Met': 0, 'Val': 0, 'Ser': 0, 'Pro': 0, 'Thr': 0, 'Ala': 0, 'Tyr': 0, 'His': 0, 'Gln': 0, 'Asn': 0, 'Lys': 0, 'Asp': 0, 'Glu': 0, 'Cys': 0, 'Trp': 0, 'Arg': 0, 'Gly': 0}\n",
    "        for aa in self.aa_seq:\n",
    "            for prot in sum_dict:\n",
    "                if aa == prot:\n",
    "                    sum_dict[prot] += 1\n",
    "        \n",
    "        print(f\"This protein has {self.aa_count} amino acids.\")\n",
    "        for key, value in sum_dict.items():\n",
    "            print(f\"{key:>30}:{value}\")\n",
    "    \n",
    "    #writes protein into a file\n",
    "    def write_protein(self):\n",
    "        f = open(str(self.prot_num),\"w\")\n",
    "        f.write(str(self.aa_seq))\n",
    "        f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Sequence:\n",
    "    def __init__(self, reference, source, sub_date, acc_date, coding_type, na_seq):\n",
    "        self.reference = reference\n",
    "        self.source = source\n",
    "        self.sub_date = sub_date\n",
    "        self.acc_date = acc_date\n",
    "        self.coding_type = coding_type\n",
    "        self.na_seq = na_seq\n",
    "        self.orf_list = []\n",
    "    \n",
    "    #Number of basepairs in a sequence\n",
    "    def basepairs(self):\n",
    "        return len(self.na_seq)\n",
    "    \n",
    "    def print_attributes(self):\n",
    "        print(self.reference,self.source,self.sub_date,self.acc_date,self.coding_type,self.na_seq,self.orf_list)\n",
    "    \n",
    "    #Finds the different orfs in the sequence, converts into ORF object, and stores in a list\n",
    "    def find_orfs(self):     \n",
    "        print(self.coding_type)\n",
    "        if self.coding_type == \"RNA\": #Turn RNA into DNA\n",
    "            self.na_seq = self.na_seq.replace(\"U\",\"T\")\n",
    "        \n",
    "        #variables for the ORF object\n",
    "        orf_num = 0\n",
    "        orf_bps = 0\n",
    "        bp_start = 0\n",
    "        bp_stop = 0\n",
    "        orf_seq = []\n",
    "        \n",
    "        started = False\n",
    "        i = 0\n",
    "        #while loop goes through whole sequence and pulls out Orfs.\n",
    "        while i < len(self.na_seq)-2:\n",
    "            triple_base = self.na_seq[i:i+3]\n",
    "            \n",
    "            if started == False and triple_base == \"ATG\": #starting the orf reading\n",
    "                started = True\n",
    "                bp_start = i + 1\n",
    "            elif started == True and (triple_base == \"TGA\" or triple_base == \"TAA\" or triple_base == \"TAG\"): #ending the orf\n",
    "                started = False\n",
    "                \n",
    "                #calculating orf variables\n",
    "                orf_num += 1\n",
    "                bp_stop = i + 3\n",
    "                orf_bps = bp_stop - bp_start + 1\n",
    "                \n",
    "                an_orf = ORF(orf_num,orf_bps,bp_start,bp_stop,orf_seq)#create orf object\n",
    "                self.orf_list.append(an_orf)\n",
    "                \n",
    "                orf_seq = []#renewing the sequence list for the next orf\n",
    "            \n",
    "            #different increments depending on if currently reading an orf\n",
    "            if started:\n",
    "                orf_seq.append(triple_base)\n",
    "                i += 3 #skip next 2 iterations because I stored this codon(3 base pairs)\n",
    "            else:\n",
    "                i += 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29902\n",
      "DNA\n",
      "11 total proteins in orf_list\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 1\n",
      "Basepairs: 21291\n",
      "Starting at 265, ending at 21555\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 2\n",
      "Basepairs: 3822\n",
      "Starting at 21562, ending at 25383\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 3\n",
      "Basepairs: 831\n",
      "Starting at 25390, ending at 26220\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 4\n",
      "Basepairs: 228\n",
      "Starting at 26245, ending at 26472\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 5\n",
      "Basepairs: 669\n",
      "Starting at 26521, ending at 27189\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 6\n",
      "Basepairs: 186\n",
      "Starting at 27202, ending at 27387\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 7\n",
      "Basepairs: 366\n",
      "Starting at 27394, ending at 27759\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 8\n",
      "Basepairs: 123\n",
      "Starting at 27766, ending at 27888\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 9\n",
      "Basepairs: 366\n",
      "Starting at 27892, ending at 28257\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 10\n",
      "Basepairs: 1260\n",
      "Starting at 28273, ending at 29532\n",
      "\n",
      "Orf attributes:\n",
      "The orf number is: 11\n",
      "Basepairs: 117\n",
      "Starting at 29557, ending at 29673\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Main code\n",
    "#reads sequence file\n",
    "f = open(\"COVID-19 Genome.txt\",\"r\")\n",
    "input = [f.readline().strip() for i in range(6)]\n",
    "a,b,c,d,e,f = input\n",
    "seq = Sequence(a,b,c,d,e,f)\n",
    "\n",
    "print(seq.basepairs())\n",
    "seq.find_orfs()\n",
    "\n",
    "print(f\"{len(seq.orf_list)} total proteins in orf_list\\n\")\n",
    "\n",
    "for i in seq.orf_list:\n",
    "    i.print_attributes()\n",
    "    #i.transcribe().print_aa()\n",
    "\n",
    "\n",
    "seq.orf_list[1].transcribe().write_protein()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
